File: editor/en/[358]Rearrange String k Distance Apart.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public String rearrangeString(String s, int k) {
        if (k <= 1) {
            return s;
        }
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a, b) -> b[1] - a[1]);
        for (int i = 0; i < 26; i++) {
            if (count[i] > 0) {
                maxHeap.offer(new int[]{i, count[i]});
            }
        }
        StringBuilder sb = new StringBuilder();
        Queue<int[]> waitQueue = new LinkedList<>();
        while (!maxHeap.isEmpty()) {
            int[] curr = maxHeap.poll();
            sb.append((char) (curr[0] + 'a'));
            curr[1]--;
            waitQueue.offer(curr);
            if (waitQueue.size() < k) {
                continue;
            }
            int[] front = waitQueue.poll();
            if (front[1] > 0) {
                maxHeap.offer(front);
            }
        }
        return sb.length() == s.length() ? sb.toString() : "";
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[358]Rearrange String k Distance Apart.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var rearrangeString = function (s, k) {
    if (k <= 1) {
        return s;
    }

    const map = new Map();

    for (const c of s) {
        map.set(c, (map.get(c) || 0) + 1);
    }

    const maxHeap = Array.from(map.entries()).sort((a, b) => b[1] - a[1]);
    const res = [];
    const waitQueue = [];

    while (maxHeap.length > 0) {
        const curr = maxHeap.shift();
        res.push(curr[0]);
        curr[1]--;
        waitQueue.push(curr);

        if (waitQueue.length < k) {
            continue;
        }

        const front = waitQueue.shift();

        if (front[1] > 0) {
            maxHeap.push(front);
            maxHeap.sort((a, b) => b[1] - a[1]);
        }
    }

    return res.length === s.length ? res.join('') : '';
};
//leetcode submit region end(Prohibit modification and deletion)
/*
var rearrangeString = function (s, k) {
    if (k <= 1) {
        return s;
    }
    const map = new Map();
    for (const c of s) {
        map.set(c, (map.get(c) || 0) + 1);
    }
    const maxHeap = Array.from(map.entries()).sort((a, b) => b[1] - a[1]);
    const res = [];
    const waitQueue = [];
    while (maxHeap.length > 0) {
        const curr = maxHeap.shift();
        res.push(curr[0]);
        curr[1]--;
        waitQueue.push(curr);
        if (waitQueue.length < k) {
            continue;
        }
        const front = waitQueue.shift();
        if (front[1] > 0) {
            maxHeap.push(front);
            maxHeap.sort((a, b) => b[1] - a[1]);
        }
    }
    return res.length === s.length ? res.join('') : '';
};

-101 ms 較慢
var rearrangeString = function (s, k) {
    if (k <= 1) {
        return s;
    }
    const map = new Map();
    for (const c of s) {
        map.set(c, (map.get(c) || 0) + 1);
    }
    const compareFunction = (a, b) => b[1] - a[1];
    const maxHeap = new MaxPriorityQueue({compare: compareFunction});
    for (const [key, value] of map) {
        maxHeap.enqueue([key, value]);
    }
    const res = [];
    const waitQueue = [];
    while (!maxHeap.isEmpty()) {
        const curr = maxHeap.dequeue();
        res.push(curr[0]);
        curr[1]--;
        waitQueue.push(curr);
        if (waitQueue.length < k) {
            continue;
        }
        const front = waitQueue.shift();
        if (front[1] > 0) {
            maxHeap.enqueue(front);
        }
    }
    return res.length === s.length ? res.join('') : '';
};
 */